diff --git a/libraries/AP_HAL_PX4/HAL_PX4_Class.cpp b/libraries/AP_HAL_PX4/HAL_PX4_Class.cpp
index 7309a61..d95cf3a 100644
--- a/libraries/AP_HAL_PX4/HAL_PX4_Class.cpp
+++ b/libraries/AP_HAL_PX4/HAL_PX4_Class.cpp
@@ -44,7 +44,8 @@ static PX4Util utilInstance;
 static PX4GPIO gpioDriver;
 
 #if defined(CONFIG_ARCH_BOARD_PX4FMU_V2)
-#define UARTA_DEFAULT_DEVICE "/dev/ttyACM0"
+//#define UARTA_DEFAULT_DEVICE "/dev/ttyACM0"
+#define UARTA_DEFAULT_DEVICE "/dev/ttyS0"
 #define UARTB_DEFAULT_DEVICE "/dev/ttyS3"
 #define UARTC_DEFAULT_DEVICE "/dev/ttyS1"
 #define UARTD_DEFAULT_DEVICE "/dev/ttyS2"
diff --git a/libraries/AP_HAL_PX4/NSHShellStream.cpp b/libraries/AP_HAL_PX4/NSHShellStream.cpp
index a8a6bb1..9c1c0bf 100644
--- a/libraries/AP_HAL_PX4/NSHShellStream.cpp
+++ b/libraries/AP_HAL_PX4/NSHShellStream.cpp
@@ -12,7 +12,7 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <apps/nsh.h>
+#include <apps/nshlib/nshlib.h>
 #include <fcntl.h>
 #include <sys/ioctl.h>
 #include "Scheduler.h"
diff --git a/libraries/AP_HAL_PX4/Util.cpp b/libraries/AP_HAL_PX4/Util.cpp
index ba0c4f2..e9c83b7 100644
--- a/libraries/AP_HAL_PX4/Util.cpp
+++ b/libraries/AP_HAL_PX4/Util.cpp
@@ -6,7 +6,7 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <apps/nsh.h>
+#include <apps/nshlib/nshlib.h>
 #include <fcntl.h>
 #include "UARTDriver.h"
 #include <uORB/uORB.h>
diff --git a/mk/px4_targets.mk b/mk/px4_targets.mk
index 1eb25ee..d3c1ab8 100644
--- a/mk/px4_targets.mk
+++ b/mk/px4_targets.mk
@@ -43,6 +43,7 @@ EXTRAFLAGS += -DUAVCAN=1
 # we have different config files for V1 and V2
 PX4_V1_CONFIG_FILE=$(MK_DIR)/PX4/config_px4fmu-v1_APM.mk
 PX4_V2_CONFIG_FILE=$(MK_DIR)/PX4/config_px4fmu-v2_APM.mk
+PX4_NIU_CONFIG_FILE=$(MK_DIR)/PX4/config_px4fmu-niu_APM.mk
 
 SKETCHFLAGS=$(SKETCHLIBINCLUDES) -DARDUPILOT_BUILD -DTESTS_MATHLIB_DISABLE -DCONFIG_HAL_BOARD=HAL_BOARD_PX4 -DSKETCHNAME="\\\"$(SKETCH)\\\"" -DSKETCH_MAIN=ArduPilot_main -DAPM_BUILD_DIRECTORY=APM_BUILD_$(SKETCH)
 
@@ -99,13 +100,28 @@ px4-v2: $(BUILDROOT)/make.flags CHECK_MODULES $(PX4_ROOT)/Archives/px4fmu-v2.exp
 	$(v) $(SKETCHBOOK)/Tools/scripts/add_git_hashes.py $(HASHADDER_FLAGS) "$(SKETCH)-v2.px4" "$(SKETCH)-v2.px4"
 	$(v) echo "PX4 $(SKETCH) Firmware is in $(SKETCH)-v2.px4"
 
+_px4-niu: $(PX4_ROOT)/Archives/px4fmu-niu.export
+px4-niu: $(BUILDROOT)/make.flags CHECK_MODULES $(PX4_ROOT)/Archives/px4fmu-niu.export $(SKETCHCPP) module_mk
+	$(RULEHDR)
+	$(v) rm -f $(PX4_ROOT)/makefiles/$(PX4_NIU_CONFIG_FILE)
+	$(v) cp $(PX4_NIU_CONFIG_FILE) $(PX4_ROOT)/makefiles/nuttx/
+	$(PX4_MAKE) px4fmu-niu_APM
+	$(v) /bin/rm -f $(SKETCH)-niu.px4
+	$(v) cp $(PX4_ROOT)/Images/px4fmu-niu_APM.px4 $(SKETCH)-niu.px4
+	$(v) $(SKETCHBOOK)/Tools/scripts/add_git_hashes.py $(HASHADDER_FLAGS) "$(SKETCH)-niu.px4" "$(SKETCH)-niu.px4"
+	$(v) echo "PX4 $(SKETCH) Firmware is in $(SKETCH)-niu.px4"
+
 px4: px4-v1 px4-v2
 
 px4-clean: clean CHECK_MODULES px4-archives-clean px4-cleandep
-	$(v) /bin/rm -rf $(PX4_ROOT)/makefiles/build $(PX4_ROOT)/Build
+#	$(v) /bin/rm -rf $(PX4_ROOT)/makefiles/build $(PX4_ROOT)/Build
+	$(v) /bin/rm -rf $(PX4_ROOT)/makefiles/build $(PX4_ROOT)/Build $(PX4_ROOT)/Images/*.px4 $(PX4_ROOT)/Images/*.bin
+	$(v) /bin/rm -rf $(PX4_ROOT)/src/modules/uORB/topics $(PX4_ROOT)/src/platforms/nuttx/px4_messages
 
 px4-cleandep: clean
+	$(v) mkdir -p $(PX4_ROOT)/Build
 	$(v) find $(PX4_ROOT)/Build -type f -name '*.d' | xargs rm -f
+	$(v) find $(UAVCAN_DIRECTORY) -type f -name '*.d' | xargs rm -f
 	$(v) find $(SKETCHBOOK)/$(SKETCH) -type f -name '*.d' | xargs rm -f
 
 px4-v1-upload: px4-v1
@@ -158,6 +174,9 @@ $(PX4_ROOT)/Archives/px4fmu-v1.export:
 $(PX4_ROOT)/Archives/px4fmu-v2.export:
 	$(v) $(PX4_MAKE_ARCHIVES) BOARDS="px4fmu-v2"
 
+$(PX4_ROOT)/Archives/px4fmu-niu.export:
+	$(v) $(PX4_MAKE_ARCHIVES) BOARDS="px4fmu-niu"
+
 $(PX4_ROOT)/Archives/px4io-v1.export:
 	$(v) $(PX4_MAKE_ARCHIVES) BOARDS="px4io-v1"
 
diff --git a/mk/targets.mk b/mk/targets.mk
index 5b0a64a..164b3b8 100644
--- a/mk/targets.mk
+++ b/mk/targets.mk
@@ -77,7 +77,7 @@ empty: all
 
 # cope with copter and hil targets
 FRAMES = quad tri hexa y6 octa octa-quad heli single coax obc nologging
-BOARDS = apm1 apm2 apm2beta apm1-1280 px4 px4-v1 px4-v2 sitl flymaple linux vrbrain vrbrain-v40 vrbrain-v45 vrbrainv-50 vrbrain-v51 vrbrain-v52 vrubrain-v51 vrubrain-v52 vrhero-v10 erle pxf navio bbbmini
+BOARDS = apm1 apm2 apm2beta apm1-1280 px4 px4-v1 px4-v2 sitl flymaple linux vrbrain vrbrain-v40 vrbrain-v45 vrbrainv-50 vrbrain-v51 vrbrain-v52 vrubrain-v51 vrubrain-v52 vrhero-v10 erle pxf navio bbbmini px4-niu
 
 define frame_template
 $(1)-$(2) : EXTRAFLAGS += "-DFRAME_CONFIG=$(shell echo $(2) | tr a-z A-Z | sed s/-/_/g)_FRAME "
diff --git a/modules/PX4Firmware b/modules/PX4Firmware
--- a/modules/PX4Firmware
+++ b/modules/PX4Firmware
@@ -1 +1 @@
-Subproject commit 34e1d543e49c2756b8fa4b81ef30468497ea792a
+Subproject commit 34e1d543e49c2756b8fa4b81ef30468497ea792a-dirty
diff --git a/modules/PX4NuttX b/modules/PX4NuttX
--- a/modules/PX4NuttX
+++ b/modules/PX4NuttX
@@ -1 +1 @@
-Subproject commit 7c5ef883fc3307b44bef9d5f7b51cb144cdf400a
+Subproject commit 7c5ef883fc3307b44bef9d5f7b51cb144cdf400a-dirty
diff --git a/modules/uavcan b/modules/uavcan
--- a/modules/uavcan
+++ b/modules/uavcan
@@ -1 +1 @@
-Subproject commit 6dd432c9742c22e1dd1638c7f91cf937e4bdb2f1
+Subproject commit 6dd432c9742c22e1dd1638c7f91cf937e4bdb2f1-dirty
